h/accounts/test/views_test.py

```diff
@@ -48,228 +46,275 @@ def configure(config):
+    result = ajax_form(request, {'errors': 'data'})
+
+    assert result['status'] == 'failure'
+
+
+def test_ajax_form_sets_status_code_400_on_errors():
+    request = DummyRequest()
+    result = ajax_form(request, {'errors': 'data'})
+
+    assert request.response.status_code == 400
+
+
+def test_ajax_form_sets_status_code_from_input_on_errors():
+    request = DummyRequest()
+    result = ajax_form(request, {'errors': 'data', 'code': 418})
+
+    assert request.response.status_code == 418
+
+
+def test_ajax_form_aggregates_errors_on_success():
+    request = DummyRequest()
+    errors = [
+        {'name': 'Name is too weird'},
+        {'email': 'Email must be @hotmail.com'},
+    ]
+    result = ajax_form(request, {'errors': errors})
+
+    assert result['errors'] == {'name': 'Name is too weird',
+                                'email': 'Email must be @hotmail.com'}
+
+
+def test_ajax_form_passes_data_through_on_success():
+    request = DummyRequest()
+    result = ajax_form(request, {'some': 'data', 'no': 'errors'})
+
+    assert result['some'] == 'data'
+    assert result['no'] == 'errors'
+    assert request.response.status_code == 200
+
+
+def test_ajax_form_ignores_status_code_from_input_on_success():
+    request = DummyRequest()
+    result = ajax_form(request, {'some': 'data', 'code': 418})
+
+    assert request.response.status_code == 200
+
+
+def test_ajax_form_includes_flash_data(pop_flash):
+    request = DummyRequest()
+    pop_flash.return_value = {'success': ['Well done!']}
+    result = ajax_form(request, {'some': 'data'})
+
+    assert result['flash'] == {'success': ['Well done!']}
+
+
+def test_ajax_form_sets_status_code_400_on_flash_error(pop_flash):
+    request = DummyRequest()
+    pop_flash.return_value = {'error': ['I asplode!']}
+    result = ajax_form(request, {'some': 'data'})
+
+    assert request.response.status_code == 400
+
+
+def test_ajax_form_sets_status_failure_on_flash_error(pop_flash):
+    request = DummyRequest()
+    pop_flash.return_value = {'error': ['I asplode!']}
+    result = ajax_form(request, {'some': 'data'})
+
+    assert result['status'] == 'failure'
+
 
-    def get_fake_token():
-        return 'fake_token'
+def test_ajax_form_sets_reason_on_flash_error(pop_flash):
+    request = DummyRequest()
+    pop_flash.return_value = {'error': ['I asplode!']}
+    result = ajax_form(request, {'some': 'data'})
 
-    fake_request.params['csrf_token'] = 'fake_token'
-    fake_request.session.get_csrf_token = get_fake_token
-    fake_request.POST['username'] = username
-    fake_request.POST['pwd'] = password
+    assert result['reason'] == 'I asplode!'
 
-    if with_subscriptions:
-        subs = '{"active": activestate, "uri": "username", "id": 1}'
-        subs = subs.replace('activestate', str(active).lower()).replace('username', username)
-        fake_request.POST['subscriptions'] = subs
-    return fake_request
 
+def test_validate_form_passes_data_to_validate():
+    idata = {}
```

Comment:
If we're looking for better names for these things I would suggest maybe the ones that colander and deform use:

cstruct and appstruct. So cstruct is short for "colander struct". http://docs.pylonsproject.org/projects/colander/en/latest/glossary.html#glossary

The idea is that your form sends a formencoded sequence that is a colander struct, and colander deserializes this into an appstruct, or application structure.

In our case, with AJAX and JSON, this isn't actually a cstruct, probably, since we're skipping any deserialization colander might do, afaik, since part of the purpose of colander is to take a sequence of formencoded values submitted by a dumb, unscripted form, and turn that into a hierarchical structure that represents the logical form rather than a sequence of input controls.

<shrug>

So, it might be `in_appstruct`. Honestly, I'm not picky about the tests here, though.